common/scala/src/main/scala/whisk/core/containerpool/logging/ElasticSearchLogStore.scala

```diff
@@ -0,0 +1,130 @@
+}
+
+case class ElasticSearchLogStoreConfig(protocol: String,
+                                       host: String,
+                                       port: Int,
+                                       logMessageField: String,
+                                       activationIdField: String,
+                                       requiredHeaders: String)
+
+/**
+ * Docker based implementation of a LogStore.
+ *
+ * Relies on docker's implementation details with regards to the JSON log-driver. When using the JSON log-driver
+ * docker writes stdout/stderr to a JSON formatted file which is read by this store. Logs are written in the
+ * activation record itself.
+ *
+ * Additionally writes logs to a separate file which can be processed by any backend service asynchronously.
+ */
+class ElasticSearchLogStore(
+  system: ActorSystem,
+  httpFlow: Option[Flow[(HttpRequest, Promise[HttpResponse]), (Try[HttpResponse], Promise[HttpResponse]), Any]] = None,
+  destinationDirectory: Path = Paths.get("logs"),
+  elasticSearchConfig: ElasticSearchLogStoreConfig =
+    loadConfigOrThrow[ElasticSearchLogStoreConfig](ConfigKeys.elasticSearch))
+    extends DockerToActivationFileLogStore(system, destinationDirectory) {
+
+  implicit val actorSystem = system
+  implicit val logging = new AkkaLogging(system.log)
+
+  val protocol = elasticSearchConfig.protocol
+  val host = elasticSearchConfig.host
+  val port = elasticSearchConfig.port
+  val logMessageField = elasticSearchConfig.logMessageField
+  val activationIdField = elasticSearchConfig.activationIdField
+  val requiredHeaders = elasticSearchConfig.requiredHeaders.split(",")
+  val esClient = new ElasticSearchRestClient(protocol, host, port, httpFlow)
+
+  private def transcribeLogs(queryResult: EsSearchResult): ActivationLogs = {
+    val logs = queryResult.hits.hits.map(hit => {
+      val userLogEntry = hit.source.convertTo[UserLogEntry]
+      s"${userLogEntry.time} ${userLogEntry.stream}: ${userLogEntry.message.stripLineEnd}"
+    })
+
+    ActivationLogs(logs)
+  }
+
+  private def extractRequiredHeaders(request: HttpRequest) =
+    request.headers.filter {
+      case header: HttpHeader if requiredHeaders.contains(header.lowercaseName) => true
+      case _                                                                    => false
+    }.toList
+
+  override def fetchLogs(user: Identity, activation: WhiskActivation, request: HttpRequest): Future[ActivationLogs] = {
+    val headers = extractRequiredHeaders(request)
+
+    if (headers.length == requiredHeaders.length) {
+      val date = LocalDate.now.format(DateTimeFormatter.ofPattern("yyyy.MM.dd"))
+      val path = Uri(s"/elasticsearch/logstash-${user.uuid.asString}-$date/_search")
+      val logQuery = s"_type: $logMessageField AND $activationIdField: ${activation.activationId}"
+      val queryString = EsQueryString(EsQueryValue(logQuery))
+      val sort = Some(Array(EsQueryTimestamp(EsQueryOrder("asc")).toJson))
```

Comment:
Would be nice to not have to call `toJson` for the arguments here and on the line below.